maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java

```diff
@@ -0,0 +1,122 @@
+        if (!spaProperties.isEmpty()) {
+            JSONObject spaConfigJson = convertPropertiesToJSON(spaProperties);
+            File spaConfigFile = new File(serverDir, "microfrontends.json");
+            writeJSONObject(spaConfigFile, spaConfigJson);
+            nodeHelper.installNodeAndNpm(NODE_VERSION, NPM_VERSION);
+            File buildTargetDir = new File(serverDir, BUILD_TARGET_DIR);
+            nodeHelper.runNpx(String.format("openmrs@latest build --target %s", buildTargetDir));
+            nodeHelper.runNpx(String.format("openmrs@latest assemble --target %s --mode config --config %s", buildTargetDir, spaConfigFile));
+        }
+    }
+
+    private JSONObject convertPropertiesToJSON(Map<String, String> properties) throws RuntimeException {
+        Set<String> foundPropertySetKeys = new HashSet<>();
+        JSONObject result = new JSONObject();
+        for (String dotDelimitedKeys : properties.keySet()) {
+            String[] keys = dotDelimitedKeys.split("\\.");
+            for (int i = 0; i < keys.length - 1; i++) {
+                foundPropertySetKeys.add(StringUtils.join(Arrays.copyOfRange(keys, 0, i), "."));
+            }
+        }
+        for (String dotDelimitedKeys : properties.keySet()) {
+            if (foundPropertySetKeys.contains(dotDelimitedKeys)) {
+                String badLine = "spa." + dotDelimitedKeys + "=" + properties.get(dotDelimitedKeys);
+                throw new RuntimeException(BAD_SPA_PROPERTIES_MESSAGE +
+                        " The following property is a parent property to another, and therefore cannot be assigned a value:\t\"" + badLine + "\"");
+            }
+            addPropertyToJSONObject(result, dotDelimitedKeys, properties.get(dotDelimitedKeys));
+        }
+        return result;
+    }
+
+    private void addPropertyToJSONObject(JSONObject jsonObject, String propertyKey, String value) {
+        String[] keys = propertyKey.split("\\.");
+        if (keys.length == 1) {
+            if (jsonObject.containsKey(keys[0])) {
+                throw new RuntimeException(BAD_SPA_PROPERTIES_MESSAGE +
+                        " Encountered this error processing a property containing the key '" + keys[0] + "' and with value " + value);
+            }
+            jsonObject.put(keys[0], value);
+        } else {
+            if (!jsonObject.containsKey(keys[0])) {
+                jsonObject.put(keys[0], new JSONObject());
+            }
+            Object childObject = jsonObject.get(keys[0]);
+            if (!(childObject instanceof JSONObject)) {
+                throw new RuntimeException(BAD_SPA_PROPERTIES_MESSAGE +
+                        " Also please post to OpenMRS Talk and include this full message. If you are seeing this, there has been a programming error.");
+            }
+            String childKeys = StringUtils.join(Arrays.copyOfRange(keys, 1, keys.length), ".");
```

Comment:
```suggestion
            String childKeys = propertyKey.substring(propertyKey.indexOf(".") + 1)
```

Reply:
Any reason for this? The StringUtils version will be easier to refactor, since it leaves the only reference to `propertyKey` in the function in the first line. They seem about equally simple to me.

Reply:
Mostly, just for less data copying, but for the expected one-value nest, I don't think there's much of a difference really.